url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/classloading/CascadingClassLoadHelper.java

```diff
@@ -0,0 +1,310 @@
+    }
+    logger.info("Package: '" + pkgname + "' becomes Resource: '" + resource.toString() + "'");
+
+    if (resource.toString().startsWith("jar:")) {
+      processJarfile(resource, pkgname, classes);
+    } else {
+      processDirectory(new File(resource.getPath()), pkgname, classes);
+    }
+
+    return classes;
+
+  }
+
+  private void processDirectory(File directory, String pkgname, Set<Class<? extends Job>> classes) {
+
+    logger.debug("Reading Directory '" + directory + "'");
+    // Get the list of the files contained in the package
+    String[] files = directory.list();
+    for (int i = 0; i < files.length; i++) {
+      String fileName = files[i];
+      String className = null;
+      // we are only interested in .class files
+      if (fileName.endsWith(".class")) {
+        // removes the .class extension
+        className = pkgname + '.' + fileName.substring(0, fileName.length() - 6);
+      }
+      logger.debug("FileName '" + fileName + "'  =>  class '" + className + "'");
+      if (className != null) {
+        filterJobClassWithExceptionCatch(className, classes);
+      }
+      File subdir = new File(directory, fileName);
+      if (subdir.isDirectory()) {
+        processDirectory(subdir, pkgname + '.' + fileName, classes);
+      }
+    }
+  }
+
+  private void processJarfile(URL resource, String pkgname, Set<Class<? extends Job>> classes) {
+
+    String relPath = pkgname.replace('.', '/');
+    String resPath = resource.getPath().replace("%20", " ");
+    String jarPath = resPath.replaceFirst("[.]jar[!].*", ".jar").replaceFirst("file:", "");
+    logger.debug("Reading JAR file: '" + jarPath + "'");
+    JarFile jarFile;
+    try {
+      jarFile = new JarFile(jarPath);
+    } catch (IOException e) {
+      throw new RuntimeException("Unexpected IOException reading JAR File '" + jarPath + "'", e);
+    }
+    Enumeration<JarEntry> entries = jarFile.entries();
+    while (entries.hasMoreElements()) {
+      JarEntry entry = entries.nextElement();
+      String entryName = entry.getName();
+      String className = null;
+      if (entryName.endsWith(".class") && entryName.startsWith(relPath) && entryName.length() > (relPath.length() + "/".length())) {
+        className = entryName.replace('/', '.').replace('\\', '.').replace(".class", "");
+      }
+      logger.debug("JarEntry '" + entryName + "'  =>  class '" + className + "'");
+      if (className != null) {
+        filterJobClassWithExceptionCatch(className, classes);
+      }
+    }
+  }
+
+  private void filterJobClassWithExceptionCatch(String className, Set<Class<? extends Job>> classes) {
+    try {
+
+      Class clazz = loadClass(className);
+      if (Modifier.isAbstract(clazz.getModifiers())) {
+          return;
+      }
+      if (Modifier.isInterface(clazz.getModifiers())) {
+          return;
+      }
+      if (Job.class.isAssignableFrom(clazz)) {
+        classes.add(clazz);
+      }
+    } catch (ClassNotFoundException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)